File: java/GameSimulator.java
package com.example.game;

import io.opentelemetry.api.common.Attributes;
import io.opentelemetry.api.metrics.LongCounter;
import io.opentelemetry.api.metrics.DoubleHistogram;
import io.opentelemetry.api.metrics.Meter;
import io.opentelemetry.exporter.otlp.metrics.OtlpGrpcMetricExporter;
import io.opentelemetry.sdk.metrics.SdkMeterProvider;
import io.opentelemetry.sdk.metrics.export.PeriodicMetricReader;
import io.opentelemetry.sdk.resources.Resource;
import org.slf4j.Logger;
import org.slf4j.LoggerFactory;

import java.util.Random;
import java.util.concurrent.ExecutorService;
import java.util.concurrent.Executors;
import java.util.concurrent.TimeUnit;
import java.util.concurrent.atomic.AtomicInteger;

public class GameSimulator {
    private static final Logger logger = LoggerFactory.getLogger(GameSimulator.class);
    private static final Random random = new Random();
    
    private final LongCounter gamesCounter;
    private final LongCounter guessCounter;
    private final DoubleHistogram guessDistance;
    private final AtomicInteger currentPlayers = new AtomicInteger(0);

    public GameSimulator() {
        Resource resource = Resource.getDefault()
            .toBuilder()
            .put("service.name", "game_simulator")
            .build();

        OtlpGrpcMetricExporter metricExporter = OtlpGrpcMetricExporter.builder()
            .setEndpoint("http://otel-collector.tme-obx.svc.cluster.local:4317")
            .build();

        SdkMeterProvider meterProvider = SdkMeterProvider.builder()
            .setResource(resource)
            .registerMetricReader(PeriodicMetricReader.builder(metricExporter)
                .setInterval(1, TimeUnit.SECONDS)
                .build())
            .build();

        Meter meter = meterProvider.get("game_simulator");

        gamesCounter = meter.counterBuilder("games_total")
            .setDescription("Total number of games played")
            .build();

        guessCounter = meter.counterBuilder("guesses_total")
            .setDescription("Total number of guesses made")
            .build();

        guessDistance = meter.histogramBuilder("guess_distance")
            .setDescription("Distance between guess and target number")
            .build();
    }

    private void simulateGame() {
        String gameId = String.valueOf(System.currentTimeMillis());
        int target = random.nextInt(20) + 1;
        int attempts = 0;
        int currentGuess;

        logger.info("Starting new game with ID: {}", gameId);

        do {
            attempts++;
            currentGuess = random.nextInt(20) + 1;
            int distance = Math.abs(currentGuess - target);

            guessCounter.add(1, Attributes.builder()
                .put("game_id", gameId)
                .build());
                
            guessDistance.record(distance, Attributes.builder()
                .put("game_id", gameId)
                .build());

            try {
                Thread.sleep(random.nextInt(1000) + 500);
            } catch (InterruptedException e) {
                Thread.currentThread().interrupt();
                return;
            }
        } while (currentGuess != target);

        gamesCounter.add(1, Attributes.builder()
            .put("game_id", gameId)
            .put("attempts", attempts)
            .build());

        logger.info("Game {} completed in {} attempts", gameId, attempts);
    }

    public void startSimulation() {
        ExecutorService executor = Executors.newFixedThreadPool(5);
        
        while (true) {
            try {
                int numPlayers = random.nextInt(5) + 1;
                currentPlayers.set(numPlayers);
                
                logger.info("Starting batch with {} players", numPlayers);
                
                for (int i = 0; i < numPlayers; i++) {
                    executor.submit(this::simulateGame);
                }

                Thread.sleep(2000);
            } catch (InterruptedException e) {
                Thread.currentThread().interrupt();
                break;
            }
        }

        executor.shutdown();
    }

    public static void main(String[] args) {
        GameSimulator simulator = new GameSimulator();
        simulator.startSimulation();
    }
}

File: cpp/game_simulator.cpp
#include <opentelemetry/sdk/metrics/meter.h>
#include <opentelemetry/sdk/metrics/meter_provider.h>
#include <opentelemetry/exporters/otlp/otlp_grpc_metric_exporter.h>
#include <opentelemetry/sdk/metrics/periodic_exporting_metric_reader.h>
#include <opentelemetry/sdk/resource/resource.h>

#include <chrono>
#include <random>
#include <thread>
#include <iostream>

using namespace opentelemetry::sdk::metrics;
using namespace opentelemetry::sdk::resource;
namespace metrics_api = opentelemetry::metrics;

int main() {
    // Configure exporter
    opentelemetry::exporter::otlp::OtlpGrpcMetricExporterOptions opts;
    opts.endpoint = "http://otel-collector.tme-obx.svc.cluster.local:4317";
    auto exporter = std::unique_ptr<opentelemetry::sdk::metrics::PushMetricExporter>(
        new opentelemetry::exporter::otlp::OtlpGrpcMetricExporter(opts));

    // Create MeterProvider with resource attributes
    Resource resource = Resource::Create({
        {"service.name", "game_simulator"},
        {"service.namespace", "tme-obx"}
    });

    auto provider = std::make_shared<MeterProvider>(
        MeterProviderOptions().SetResource(resource).SetMetricReader(
            std::make_unique<PeriodicExportingMetricReader>(
                std::move(exporter),
                PeriodicExportingMetricReaderOptions{std::chrono::seconds(1)}
            )
        )
    );

    auto meter = provider->GetMeter("game_simulator", "1.0.0");

    // Create metrics
    auto games_counter = meter->CreateCounter<int64_t>("games_total", 
        "Total number of games played", "1");
    auto guess_counter = meter->CreateCounter<int64_t>("guesses_total",
        "Total number of guesses made", "1");
    auto guess_distance = meter->CreateHistogram<double>("guess_distance",
        "Distance between guess and target number", "1");

    std::random_device rd;
    std::mt19937 gen(rd());
    std::uniform_int_distribution<> dist(1, 20);
    std::uniform_int_distribution<> sleep_dist(500, 1500);

    while (true) {
        auto game_id = std::to_string(
            std::chrono::system_clock::now().time_since_epoch().count()
        );
        int target = dist(gen);
        int attempts = 0;
        
        std::cout << "Starting new game with target: " << target << std::endl;

        while (true) {
            attempts++;
            int current_guess = dist(gen);
            double distance = std::abs(current_guess - target);

            opentelemetry::common::KeyValueIterable game_attributes{
                {{"game_id", game_id}}
            };
            guess_counter->Add(1, game_attributes);
            guess_distance->Record(distance, game_attributes);

            if (current_guess == target) {
                opentelemetry::common::KeyValueIterable win_attributes{
                    {{"game_id", game_id}, {"attempts", std::to_string(attempts)}}
                };
                games_counter->Add(1, win_attributes);
                break;
            }

            std::this_thread::sleep_for(
                std::chrono::milliseconds(sleep_dist(gen))
            );
        }

        std::cout << "Game " << game_id << " completed in " 
                  << attempts << " attempts" << std::endl;

        std::this_thread::sleep_for(std::chrono::seconds(2));
    }

    return 0;
}
